### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by adnanh** [adnanh/webhook#513] [Ideas]
Just a list of stuff that should be done / discussed, in no particular order.

### README.md enhancements
The current README.md has become a bit bloated and would greatly benefit from a re-organization, and some sections should probably be rewritten. A lot of issues pop up in relation to compilation issues, so it would be better if we put binary downloads in the first plan, and compilation hidden behind a link to docs or something similar. Maybe the README is not a good choice for this kind of stuff anyway, perhaps a real web-page would make it more flexible and easier to present this stuff?

### Expression rule
A trigger-rule that allows for custom expressions.

I've been playing around with [cel-go](https://github.com/google/cel-go) and it seems like a nice candidate for the execution env.
Aside from the integrated cel-go macros, in order to be able to deprecate the "snowflake" trigger rules, we'd have to expose functions that can calculate SHAxyz HMAC digests, IP CIDR matching, and probably few more.

Would solve: #504 #505 #389 #267 #200 

---

### Context provider command
The context provider command would be executed with the request details passed through STDIN, and it would be required to output a valid JSON string. The returned JSON object would then be accessible in trigger rules for further evaluation.

Proposal implementation: #356 

Would solve: #468 #416 #358 #336 #148

---

### DSL configuration format
The current JSON/YAML can be quite tedious to write, and read. It would be nice if we could move to a [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) that would allow us to write configuration in a more declarative fashion.
It would also be nice to have a specific section for the global configuration which at the moment is only available through command parameters. As a part of this effort, we could seize the opportunity to refactor the route matching logic as well to introduce the support for dynamic URL segments as well as regex route matches.
 
@moorereason played around with [HCL](https://github.com/hashicorp/hcl) for this.

**Comment by moorereason** [comment]
My thoughts...

### Documentation
I agree about the docs.  Needs work.  Basic markdown in Github can make it difficult to convey complex concepts.  If we make a website, I'd recommend one of the other projects I contribute to, [Hugo](https://github.com/gohugoio/hugo).  

### Expressions
I was tinkering with `github.com/antonmedv/expr`, which looks simpler than cel-go.  I haven't done much with it yet beyond a very basic smoke test.

### DSL

I would love to have HCL solve all of our problems.  :smiley:  I've spent most of my tinker time here.  My HCL proof-of-concept playground is https://github.com/moorereason/webhook-hcl.  It's not well-documented or meant for anyone else to understand at this point, but feel free to check it out.  It doesn't do much compared to webhook.  It loads the config, adds a mock request, and evaluates the results.

My latest config iteration is in [`config4.hcl`](https://github.com/moorereason/webhook-hcl/blob/235cec2a2482397e7d64fb1cb6bcc842ad0422cc/config4.hcl).  I was trying to capture _all_ existing webhook features and enhancement requests in the config (see [TODO.md](https://github.com/moorereason/webhook-hcl/blob/master/TODO.md) for the current progress...I just noticed that the markdown doesn't render well...look at the raw text).  For a "normal" config without the kitchen sink, see [config-github.hcl](https://github.com/moorereason/webhook-hcl/blob/master/config-github.hcl).

The most interesting thing I've done lately is add functions (which solves the expression problem) that can be seen mainly in the [constraints section](https://github.com/moorereason/webhook-hcl/blob/235cec2a2482397e7d64fb1cb6bcc842ad0422cc/config4.hcl#L30-L55) of the config.  It works pretty well, but it may be hard to provide feedback to the user on which condition was false or something for debug purposes.  My hunch is that we would need to walk through the constraints expressions to get that level of detail, but I haven't worked that out yet.  In order to do that, we'll probably have to get in the weeds of HCL and not be able to use the higher level API for that step.

Needs lots of polishing and work, but I think the HCL path has a lot of promise.  If you want to play along, PRs are welcome!  :wink:

**Reply by adnanh** [reply]
Oh wow, the antonmedv/expr looks fantastic and much simpler than cel-go!

**Comment by adnanh** [comment]
I bought a `webhook.download` domain, and configured it to temporarily redirect to the repository. It can be a new home for the README.md, installation instructions and docs.
